static/js/services/send-message.js

```diff
@@ -18,69 +18,123 @@ var _ = require('underscore'),
+              contact: undefined
             };
-          })
-          .filter(function(recipient) {
-            return recipient.phone;
-          })
-          .value();
+          }
+
+          recipient = recipient.doc;
+
+          if (recipient.phone) {
+            return {
+              phone: recipient.phone,
+              contact: recipient
+            };
+          } else if (recipient.contact && recipient.contact.phone) {
+            return {
+              phone: recipient.contact.phone,
+              contact: recipient
+            };
+          }
+        });
+      };
+
+      var descendants = function(recipient) {
+        return DB().query('medic-client/contacts_by_parent_name_type', {
+          include_docs: true,
+          startkey: [recipient.doc._id],
+          endkey: [recipient.doc._id, {}]
+        }).then(function(results) {
+          return results.rows;
+        });
+      };
+
+      var hydrate = function(recipients) {
+       return DB().allDocs({
+        include_docs: true,
+        keys: _.pluck(recipients, 'id')
+       }).then(function(results) {
+        return results.rows;
+       });
+      };
+
+      var resolvePhoneNumbers = function(recipients) {
+        //TODO: do we want to attempt to resolve phone numbers into existing contacts?
+        // users will have already got that suggestion in the send-message UI if
+        // it exists in the DB
+        return _.pluck(recipients, 'text');
       };
 
       var formatRecipients = function(recipients) {
-        var result = _.flatten(_.map(recipients, function(r) {
-          return mapRecipients(r.everyoneAt ? r.descendants : [ r ]);
-        }));
-        return _.uniq(result, false, function(r) {
-          return r.phone;
+        var splitRecipients = _.groupBy(recipients, function(recipient) {
+          if (recipient.everyoneAt) {
+            return 'explode';
+          } else if (recipient.doc){
+            return 'hydrate';
+          } else {
+            return 'resolve';
+          }
+        });
+
+        // TODO: do this better
+        splitRecipients.explode = splitRecipients.explode || [];
+        splitRecipients.hydrate = splitRecipients.hydrate || [];
+        splitRecipients.resolve = splitRecipients.resolve || [];
+
+        var promises = _.flatten(
+          [splitRecipients.explode.map(descendants),
+          hydrate(splitRecipients.hydrate),
+          resolvePhoneNumbers(splitRecipients.resolve)]
+        );
+
+        return $q.all(promises).then(function(recipients) {
+          return _.uniq(
+              // re: flatten; hydrate() and resolvePhoneNumbers() are promises with multiple values
+              // re: !!i, identity function, ie removes any undefined values caused by bad data
+              mapRecipients(_.flatten(recipients)).filter(function(i) { return !!i;}),
+              false, function(recipient) {
+                return recipient.phone;
+              });
```

Comment:
This is really hard to read. Can you break it up into multiple named vars or something?